cmd/cli/uninstall_mesh.go

```diff
@@ -80,70 +81,303 @@ func newUninstallMeshCmd(config *action.Configuration, in io.Reader, out io.Writ
 			}
 
+			uninstall.extensionsClientset, err = extensionsClientset.NewForConfig(kubeconfig)
+			if err != nil {
+				return errors.Errorf("Could not access extension client set: %s", err)
+			}
+
+			uninstall.meshNamespace = settings.Namespace()
 			return uninstall.run()
 		},
 	}
 
 	f := cmd.Flags()
 	f.StringVar(&uninstall.meshName, "mesh-name", defaultMeshName, "Name of the service mesh")
 	f.BoolVarP(&uninstall.force, "force", "f", false, "Attempt to uninstall the osm control plane instance without prompting for confirmation.  If the control plane with specified mesh name does not exist, do not display a diagnostic message or modify the exit status to reflect an error.")
+	f.BoolVarP(&uninstall.deleteClusterWideResources, "delete-cluster-wide-resources", "a", false, "Cluster wide resources (such as osm CRDs, mutating webhook configurations, validating webhook configurations and osm secrets) are fully deleted from the cluster after control plane components are deleted.")
 	f.BoolVar(&uninstall.deleteNamespace, "delete-namespace", false, "Attempt to delete the namespace after control plane components are deleted")
 	f.Uint16VarP(&uninstall.localPort, "local-port", "p", constants.OSMHTTPServerPort, "Local port to use for port forwarding")
+	f.StringVar(&uninstall.caBundleSecretName, "ca-bundle-secret-name", constants.DefaultCABundleSecretName, "Name of the secret for the OSM CA bundle")
 
 	return cmd
 }
 
 func (d *uninstallMeshCmd) run() error {
-	meshInfoList, _ := getMeshInfoList(d.config, d.clientSet)
-	if len(meshInfoList) == 0 {
-		if d.force {
-			return nil
-		}
-		return errors.Errorf("No osm mesh found")
+	err := d.validateCLIParams()
+	if err != nil {
+		return err
 	}
 
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 
-	ns := settings.Namespace()
+	if !settings.IsManaged() {
+		if !d.force {
+			// print a list of meshes within the cluster for a better user experience
+			fmt.Fprintf(d.out, "\nList of meshes present in the cluster:\n")
+
+			listCmd := &meshListCmd{
+				out:       d.out,
+				config:    d.config,
+				clientSet: d.clientSet,
+				localPort: d.localPort,
+			}
+
+			err := listCmd.run()
+
+			// Unable to list meshes in the cluster
+			if err != nil {
+				return err
+			}
+
+			confirm, err := confirm(d.in, d.out, fmt.Sprintf("\nUninstall OSM [mesh name: %s] in namespace [%s] and/or OSM resources ?", d.meshName, d.meshNamespace), 3)
+			if !confirm || err != nil {
+				return err
+			}
+		}
+
+		_, err = d.client.Run(d.meshName)
+		if err != nil && errors.Cause(err) == helmStorage.ErrReleaseNotFound {
+			fmt.Fprintf(d.out, "No OSM control plane with mesh name [%s] found in namespace [%s]\n", d.meshName, d.meshNamespace)
+
+			if !d.deleteClusterWideResources && !d.deleteNamespace {
+				return err
+			}
+		}
+
+		if err == nil {
+			fmt.Fprintf(d.out, "OSM [mesh name: %s] in namespace [%s] uninstalled\n", d.meshName, d.meshNamespace)
+		}
+	}
```

Comment:
If the uninstall cmd is being applied to a managed environment, can you add a message to the output indicating that the mesh cannot be uninstalled?
```suggestion
	} else {
	    fmt.Fprintf(d.out, "OSM [mesh name: %s] in namespace [%s] CANNOT be uninstall in a managed environment", d.meshName, d.meshNamespace)
	}
```